### PR TITLE
Updating Qwen2 small model configs and component/model builders

### PR DESCRIPTION
**Pull request by SalmanMohammadi**
#### Context
What is the purpose of this PR? Is it to
- [ ] add a new feature
- [ ] fix a bug
- [x] update tests and/or documentation
- [ ] other (please add here)

Removing low-memory config params from 0.5B and 1.5B Qwen2 models, since they probably don't need them and they will slow training down.
Also ensure validity between `apply_lora_to_output` and `tie_word_embeddings`

Tests:

```
                         __..--.._
  .....              .--~  .....  `.
.":    "`-..  .    .' ..-'"    :". `
` `._ ` _.'`"(     `-"'`._ ' _.' '
     ~~~      `.          ~~~
              .'
             /
            (
             ^---'

```

### CODE REVIEW

**Review by joecummings — COMMENTED** [review]
Definitely agree with the approach, but eyes aren't gunna get this one by me, no matter how cool the ASCII art is. 
* Can you run 10 steps for each of these so we can get a sense of the max memory allocated and reserved?
* Can you test the batch sizes with convergence? I think as long as memory is below 24 GB, you can increase the batch size. Otherwise might be worth keeping in a smaller grad acc value.

File: recipes/configs/qwen2/0.5B_lora.yaml [hunk 1]
```diff
@@ -23,8 +23,7 @@ model:
   _component_: torchtune.models.qwen2.lora_qwen2_0_5b
   lora_attn_modules: ['q_proj', 'v_proj']
   apply_lora_to_mlp: False
-  apply_lora_to_output: False
-  lora_rank: 8
+  lora_rank: 64
```
Comment (joecummings) [inline recipes/configs/qwen2/0.5B_lora.yaml:26]: Any reason we don't just want to allow them to LoRA-ify the output, with default to False?

Comment (SalmanMohammadi) [reply recipes/configs/qwen2/0.5B_lora.yaml:26]: 0.5B and 1.5B have tied embeddings

**Comment by SalmanMohammadi** [comment]
> Definitely agree with the approach, but eyes aren't gunna get this one by me, no matter how cool the ASCII art is.
> 
>     * Can you run 10 steps for each of these so we can get a sense of the max memory allocated and reserved?
> 
>     * Can you test the batch sizes with convergence? I think as long as memory is below 24 GB, you can increase the batch size. Otherwise might be worth keeping in a smaller grad acc value.

```
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⢀⣴⠆⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⣰⠟⠁⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⣀⣠⡄
⣄⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⡀⠀⢰⠃⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠐⢀⡄⠚⢁⣠⠴⠚⠋⠉⠉⠁
⠈⠑⢦⡀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⡇⠀⡎⠀⠀⠀⣀⡀⡀⠀⠀⠀⠠⢀⠔⠁⣠⡶⠋⠁⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠈⠻⢤⣀⠀⠈⠓⢤⣀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⡇⠀⢇⠀⠈⠁⠀⠀⠀⠀⢸⠟⡔⢁⣤⣾⣿⣿⡶⢶⣤⣤⡀⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠉⠳⢄⡀⠀⠈⠓⢤⡀⠀⠀⠀⠀⠀⠀⣸⡆⢳⡀⢸⡄⠀⠀⠀⠀⠀⠀⡇⢀⣶⣿⡿⠛⣻⣿⣯⢾⣷⡉⣷⡄⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠉⠢⣄⠀⠀⠙⢦⡀⠀⢀⡤⠾⠎⢱⡄⢣⠀⢳⡀⠀⠀⠀⠀⠀⢠⠏⣼⠉⠛⠶⢧⣤⠼⠏⠛⠛⠉⢻⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠈⠙⠦⣀⠀⠙⠰⠋⠀⠀⠀⠀⠙⠂⢣⡈⢷⡀⠀⠀⠀⢀⡟⢰⠇⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠘⡇⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠈⠓⢦⣀⠀⠀⠀⠀⠀⠀⠀⠀⠱⣄⠻⣄⠀⠀⢸⡇⣼⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⢠⡇⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠈⠓⢤⡀⠀⠀⠀⠀⠀⠀⠈⠀⠈⢷⣄⠀⢿⣿⠆⠀⠀⠀⠀⠀⠀⠀⠀⠀⢀⡾⠁⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠙⣦⣄⠀⠀⠀⠀⠀⠀⠀⠀⠈⢹⡿⠿⢦⣀⡀⠀⠀⠀⠀⠀⣀⡴⠟⠁⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⢠⠞⣴⣟⢧⣄⠀⠀⠀⠀⠀⠀⢠⠏⠁⠀⣠⣄⣉⡉⠓⠒⠋⠉⠉⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⡴⢋⡞⠁⠈⠙⢿⠀⠀⠀⠀⠀⢀⡟⠀⠀⠀⡿⠀⠈⠉⠉⠉⢻⠆⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⢀⠞⣱⢋⡀⢀⠀⠀⢸⡀⠀⠀⠀⠀⣾⠁⠀⠀⢠⠃⠀⠀⠀⠀⠀⡟⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⣠⢏⡞⢁⠞⠀⡈⠳⡀⢰⠃⠀⠀⠀⠀⢸⡀⠀⠀⢸⠀⠀⠀⠀⠀⣼⠃⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⢀⡼⣱⠋⢠⠊⠀⡼⠁⠀⡇⠸⠀⠀⠀⠀⠀⠈⣧⠀⠀⣼⠀⠀⠀⠀⢠⡏⠀⠀⠀⠀⠀⠀⠀⠀
⠓⢄⡀⠀⠀⠀⠀⠀⠀⠀⠀⠀⣠⢏⡼⠁⣠⠋⠢⣼⠁⠀⢀⣷⣀⣀⣀⣰⣟⡳⣄⡘⣦⢀⡟⠀⠀⠀⠀⣼⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠙⢦⣀⠀⠀⠀⠀⠀⠀⣴⣣⠏⠀⢠⠃⠀⠀⠉⢧⣰⠟⢳⣤⠾⠛⣿⠛⠛⡛⡛⠛⠋⠉⠉⠉⠉⣷⣇⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠈⠳⣄⠀⠀⢠⣾⡿⠁⠀⠀⠛⢦⡀⠀⠀⠈⡟⠀⠀⠙⢶⢶⢿⠒⠚⣿⣿⠛⠛⠋⠉⠉⡟⣻⠁⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠙⢶⣿⣿⡄⠀⠀⠀⠀⠀⠹⡄⠀⢸⡇⠀⠀⠀⢸⢸⣼⠀⠀⣿⣿⠀⠀⠀⠀⠀⡇⡿⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
```

you got it. fwiw I grabbed these values from the gemma 2b configs

**Comment by SalmanMohammadi** [comment]
Repporting: maximum peak memory (GB), ~median toks/sec, over 250 steps. 
Tested on a 2080 8GB.
[WandB runs here](https://wandb.ai/salman-mohammadi/torchtune/)

| **Config**              | Active | Allocated | Reserved | toks/sec |
| ----------------------- | ------ | --------- | -------- | -------- |
| 0.5B full single device | 6.7    | 6.7       | 7.8      | 800      |
| 0.5B LoRA single device | 6      | 6         | 8        | 1200     |
| 1.5B full single device | 8      | 8         | 8        | 400      |
| 1.5B LoRA single device | 6.5    | 6.5       | 7        | 450      |

If you'd prefer to see higher throughput we can bump bsz and announce with support for 12GB.

File: recipes/configs/qwen2/0.5B_lora_single_device.yaml [hunk 2]
```diff
@@ -19,11 +19,10 @@
 # Model Arguments
 model:
   _component_: torchtune.models.qwen2.lora_qwen2_0_5b
-  lora_attn_modules: ['q_proj', 'v_proj']
+  lora_attn_modules: ['q_proj', 'k_proj', 'v_proj']
```
Comment (joecummings) [inline recipes/configs/qwen2/0.5B_lora_single_device.yaml:22]: why did you add k_proj here?

File: recipes/configs/qwen2/0.5B_lora.yaml [hunk 3]
```diff
@@ -23,8 +23,7 @@ model:
...
   lora_alpha: 16
```
Comment (joecummings) [inline recipes/configs/qwen2/0.5B_lora.yaml:27]: I think rule of thumb to start is that alpha should be half of rank?

File: recipes/configs/qwen2/1.5B_full_single_device.yaml [hunk 4]
```diff
@@ -51,12 +52,15 @@ batch_size: 2
...
+  lr: 2e-5
+
 optimizer_in_bwd: True
+
 loss:
   _component_: torch.nn.CrossEntropyLoss
+
 max_steps_per_epoch: null
-gradient_accumulation_steps: 16
+gradient_accumulation_steps: 1
```
Comment (joecummings) [inline recipes/configs/qwen2/1.5B_full_single_device.yaml:63]: What's the reasoning for bumping this down?

File: recipes/configs/qwen2/1.5B_lora_single_device.yaml [hunk 5]
```diff
@@ -19,10 +19,9 @@
 # Model Arguments
 model:
   _component_: torchtune.models.qwen2.lora_qwen2_1_5b
-  lora_attn_modules: ['q_proj', 'v_proj']
+  lora_attn_modules: ['q_proj', 'k_proj']
```
Comment (joecummings) [inline recipes/configs/qwen2/1.5B_lora_single_device.yaml:22]: Why did you change this to 'k_proj'?

File: recipes/configs/qwen2/1.5B_lora_single_device.yaml [hunk 6]
```diff
@@ -19,10 +19,9 @@
...
   apply_lora_to_mlp: False
-  apply_lora_to_output: False
-  lora_rank: 8
+  lora_rank: 64
   lora_alpha: 16
```
Comment (joecummings) [inline recipes/configs/qwen2/1.5B_lora_single_device.yaml:25]: Sam comment as before with alpha being half of rank

File: recipes/configs/qwen2/0.5B_lora_single_device.yaml (hunk 2 above)
Comment (SalmanMohammadi) [reply recipes/configs/qwen2/0.5B_lora_single_device.yaml:22]: I had memory spare since increasing batch size OOM-ed, thought it might help performance

File: recipes/configs/qwen2/1.5B_full_single_device.yaml (hunk 4 above)
Comment (SalmanMohammadi) [reply recipes/configs/qwen2/1.5B_full_single_device.yaml:63]: opt in bwd

**Review by joecummings — COMMENTED** [review]
Thanks for doing the tests, the WandB runs are extremely helpful. I think the 8GB looks reasonable, but I have a couple questions as to the minute changes you made to the configs. 

Not a blocker for merging this PR, but did you check how much compile helped for the 1.5B full single device and the 0.5 full single device?

File: recipes/configs/qwen2/1.5B_lora_single_device.yaml (hunk 5 above)
Comment (SalmanMohammadi) [reply recipes/configs/qwen2/1.5B_lora_single_device.yaml:22]: mistake!

**Comment by SalmanMohammadi** [comment]
> Not a blocker for merging this PR, but did you check how much compile helped for the 1.5B full single device and the 0.5 full single device?

Just gave it a quick try. No difference. Might make a difference on more modern hardware?

File: recipes/configs/qwen2/0.5B_lora_single_device.yaml [hunk 7]
```diff
@@ -19,11 +19,10 @@
...
   apply_lora_to_mlp: False
-  apply_lora_to_output: False
-  lora_rank: 8
-  lora_alpha: 16
+  lora_rank: 64
+  lora_alpha: 32
```
Comment (felipemello1) [inline recipes/configs/qwen2/0.5B_lora_single_device.yaml:25]: it should be the opposite :P 

alpha is double of rank. Check llama configs for example 
image

btw, I think that having LoRA on the MLP is better than adding 'k_proj'.

File: recipes/configs/qwen2/0.5B_lora_single_device.yaml [hunk 8]
```diff
@@ -39,20 +38,22 @@ checkpointer:
...
 shuffle: True
-batch_size: 2
+batch_size: 4
 
 # Optimizer and Scheduler
 optimizer:
   _component_: torch.optim.AdamW
   weight_decay: 0.01
-  lr: 3e-4
+  lr: 2e-3
```
Comment (felipemello1) [inline recipes/configs/qwen2/0.5B_lora_single_device.yaml:55]: is the change in LR is because we tested it / saw it somewhere, or is it just a guess?

File: recipes/configs/qwen2/1.5B_lora.yaml [hunk 9]
```diff
@@ -23,9 +23,8 @@ model:
   _component_: torchtune.models.qwen2.lora_qwen2_0_5b
```
Comment (felipemello1) [inline recipes/configs/qwen2/1.5B_lora.yaml:23]: i am misreading it? The config name says 1.5, but the component and instructions are 0.5

I personally dont like having "." in the middle of a file name. I wonder if it should be 1_5B

File: recipes/configs/qwen2/0.5B_lora_single_device.yaml (hunk 8 above)
Comment (SalmanMohammadi) [reply recipes/configs/qwen2/0.5B_lora_single_device.yaml:55]: I tested when benchmarking

File: recipes/configs/qwen2/1.5B_lora.yaml (hunk 9 above)
Comment (SalmanMohammadi) [reply recipes/configs/qwen2/1.5B_lora.yaml:23]: good catch! I hadn't tested distributed. I'm not married to the config name but I agree with you

File: recipes/configs/qwen2/0.5B_lora_single_device.yaml (hunk 7 above)
Comment (SalmanMohammadi) [reply recipes/configs/qwen2/0.5B_lora_single_device.yaml:25]: I don't think this is consistent across our configs. For example, Gemma2B has:

```
model:
  _component_: torchtune.models.gemma.lora_gemma_2b
  lora_attn_modules: ['q_proj', 'k_proj', 'v_proj']
  apply_lora_to_mlp: True
  lora_rank: 64
  lora_alpha: 16
```

I think you're right though. Happy to change this

Comment (SalmanMohammadi) [reply recipes/configs/qwen2/0.5B_lora_single_device.yaml:25]: note: `apply_lora_to_mlp` did impact training speed/mem usage for me without significantly improving loss

**Comment by SalmanMohammadi** [comment]
small note: I needed https://github.com/pytorch/torchtune/issues/1185 to make these experiments fit on my GPU.

**Comment by SalmanMohammadi** [comment]
> small note: I needed #1185 to make these experiments fit on my GPU.

not sure if this pushes the needle on that issue? it's listed on the docs in stable (https://pytorch.org/docs/stable/notes/cuda.html#environment-variables) but denoted "experimental"

**Comment by joecummings** [comment]
> small note: I needed #1185 to make these experiments fit on my GPU.

All experiments? Or just the 1.5 full finetuning?

File: recipes/configs/qwen2/0.5B_lora.yaml [hunk 10]
```diff
@@ -65,20 +66,21 @@ loss:
...
   log_dir: ${output_dir}
+
 log_every_n_steps: 1
 log_peak_memory_stats: False
 
 # Environment
 device: cuda
 dtype: bf16
-enable_activation_checkpointing: False
+enable_activation_checkpointing: True
```
Comment (joecummings) [inline recipes/configs/qwen2/0.5B_lora.yaml:83]: We need activation checkpointing now?

File: recipes/configs/qwen2/0.5B_lora_single_device.yaml (hunk 7 above)
Comment (joecummings) [reply recipes/configs/qwen2/0.5B_lora_single_device.yaml:25]: Ah misspoke - definitely go with Felipe's suggestion. Alpha being double rank.

File: recipes/configs/qwen2/0.5B_lora.yaml (hunk 10 above)
Comment (SalmanMohammadi) [reply recipes/configs/qwen2/0.5B_lora.yaml:83]: I think the tradeoff for double batch size was net positive, but I can double check

**Comment by SalmanMohammadi** [comment]
> All experiments? Or just the 1.5 full finetuning?

I'm going to say all to be safe.

**Review by joecummings — APPROVED** [review]
This looks good to me - thanks for doing those tests. 

A merge on main should fix the failing CI.
